README.rst
.. image:: https://raw.githubusercontent.com/Viatorus/compile-time-printer/develop/web/static/banner-web.svg
    :alt: compile-time printer

.. image :: https://img.shields.io/github/release/viatorus/compile-time-printer.svg
    :alt: Github Releases
    :target: https://github.com/viatorus/compile-time-printer/releases
.. image :: https://img.shields.io/pypi/v/compile-time-printer.svg
    :alt: PyPI
    :target: https://pypi.org/project/compile-time-printer/
.. image :: https://api.bintray.com/packages/viatorus/compile-time-printer/Badge/images/download.svg
    :alt: Conan
    :target: https://bintray.com/viatorus/compile-time-printer/Badge/_latestVersion
.. image :: https://github.com/viatorus/compile-time-printer/workflows/testing/badge.svg
    :alt: Build Status
    :target: https://github.com/viatorus/compile-time-printer/releases
.. image :: https://img.shields.io/badge/try-online-blue.svg
    :alt: Try online
    :target: https://viatorus.github.io/compile-time-printer/

Compile-Time Printer
====================

**Compile-Time Printer** prints values and types at compile-time in C++.

Teaser
------

+-------------------------------------------------+-------------------------------------------------+
|                       test.cpp                  |    ``compile-time-printer -- make test.cpp``    |
+-------------------------------------------------+-------------------------------------------------+
| .. code-block:: cpp                             | .. code-block::                                 |
|                                                 |                                                 |
|     #include <ctp/ctp.hpp>                      |     .                                           |
|                                                 |                                                 |
|     template<auto I>                            |                                                 |
|     constexpr auto func(int i) {                |                                                 |
|                                                 |                                                 |
|       // Formatted output.                      |                                                 |
|       ctp::printf("Hello {}!\n", ctp::type{I}); |     Hello int!                                  |
|                                                 |                                                 |
|       // Standardized output.                   |                                                 |
|       ctp::print(I + i, sizeof(I));             |     42 4                                        |
|                                                 |                                                 |
|       return true;                              |                                                 |
|     }                                           |                                                 |
|                                                 |                                                 |
|     constexpr auto test = func<22>(20);         |                                               . |
+-------------------------------------------------+-------------------------------------------------+

Try it out online: https://viatorus.github.io/compile-time-printer/

Overview
--------

* `Installation`_
* `API`_
* `How it works`_
* `Known limitations`_
* `License`_

Installation
------------

Requires:

* Python >=3.6
* GCC >=7.4 and STD >=C++17

To get started, install the `python tool <https://pypi.org/project/compile-time-printer>`__:

.. code-block::

    pip install compile-time-printer

Afterwards, dump the CTP header file and include it in your project:

.. code-block::

    compile-time-printer --dump-header-file

Alternative, you can install the header file via the
`conan package <https://bintray.com/viatorus/compile-time-printer>`__.

Finally, run CTP with your build command.

*E.g. with g++ directly:*

.. code-block::

    compile-time-printer -- g++ -I. -fsyntax-only -std=c++17 -fpermissive test.cpp

*E.g. with make:*

.. code-block::

    compile-time-printer -- make test.cpp

You have to set the compiler flag *-fpermissive* in order to make it work.

API
---

ctp.hpp
+++++++

* **ctp::print(** *[file descriptor,] arguments* **)**

Prints all arguments in a simple, standardized format. Each argument is separated by one space, ending with a line
break.

.. code-block:: cpp

    int x = 42;
    ctp::print("Hello", 2.72, x);  // "Hello 2.72 42\n"

* **ctp::printf(** *[file descriptor,] format, arguments* **)**

Formats and prints all arguments in the desired format without implicit line break.
Uses the `pythons format string syntax <https://docs.python.org/3/library/string.html#format-string-syntax>`__.

.. code-block:: cpp

    int x = 42;
    ctp::printf("{:.1f}", 3.14);  // "3.1"
    ctp::printf("int: {0:d}; hex: {0:x};\n"    // "int: 42; hex: 2a;\n"
                "oct: {0:o}; bin: {0:b}", x);  // "oct: 52; bin: 101010"

* **ctp::stdout** or **ctp::stderr**

Available file descriptor to print to standard output stream or standard error stream.

.. code-block:: cpp

    ctp::print(ctp::stdout, "Info");  // stdout: "Info\n"
    ctp::printf(ctp::stderr, "Warning!\n");  // stderr: "Warning!\n"

* **ctp::type<** *Types* **>{}** or **ctp::type{** *variables* **}**

Helper object which can be passed as an argument to **print/printf** to output the type of the variables rather then
their values.

.. code-block:: cpp

    int x = 42;
    ctp::print(ctp::type<float>{}, ctp::type{x});  // "float int&\n"

* **ctp::view(** *data begin, data end* **)** or **ctp::view(** *data begin, data length* **)**
  or **ctp::view(** *contiguous range* **)** (implicit constructed)

Helper object which can be passed as an argument to **print/printf** to output a contiguous range.

.. code-block:: cpp

    int a[] = {1, 2, 3};
    ctp::print(ctp::view{a, 1}, ctp::view{a + 1, a + 3}, a);  // "[1] [2, 3] [1, 2, 3]\n"

* **ctp::formatter<** *Type* **>**

Specialize struct **ctp::formatter** for *Type*. Provide a function **constexpr auto format(** *Type* **);**
returning a tuple like object. The first element must be a format string followed by the arguments.

.. code-block:: cpp

    struct FooBar {
        int i;
    };

    template<>
    struct ctp::formatter<FooBar> {
        static constexpr auto format(const FooBar& obj) {
            return std::tuple{".i = {}", obj.i};
        }
    };

    constexpr auto test = ctp::print(FooBar{42});  // ".i = 42"

* **ctp::forward(** *value* **)** or **ctp::forward(** *function, arguments...* **)**

Helper to use **print/printf** in ``static_assert`` and template parameters. See `Known limitations`_.

compile-time-printer
++++++++++++++++++++

.. code-block::

    usage: compile-time-parser [optionals] [-- program args...]

    Compile-time printer - prints variables and types at compile time in C++.

    positional arguments:
      program               the program to compile the source (default: read from stdin)
      args                  the arguments for the command (default: [])

    optional arguments:
      -h, --help            show this help message and exit
      --version             show program's version number and exit
      -r REMOVE, --remove REMOVE
                            removes matching regex from type info (default: [])
      -cr CAPTURE_REMOVE, --capture-remove CAPTURE_REMOVE
                            removes matching regex but keeps first capture-group from type info (default: [])
      --time-point          prints time point of each print statement (default: False)
      --no-color            disables colored error output stream (default: False)
      --hide-compiler-log   don't print unparsed compiler log (default: False)
      --dump-header-file    dumps the C++ header file to ctp/ctp.hpp (default: False)

Highlights
~~~~~~~~~~

* Use ``--time-point`` to get the time when the print statement has been reached. This can be used for benchmarking.

.. code-block::

    0:00:00.236446 - Function one evaluated.
    0:00:01.238051 - Function two evaluated.

* Use ``-r`` and ``-cr`` to remove unnecessary information from types:

.. code-block:: cpp

    namespace abc::def {
        template<typename T>
        struct holder {};
    }

    using H = abc::def::holder<int>;
    constexpr auto i = ctp::print(ctp::type<H>{});  // "abc::def::holder<int>"

Output with ``-r "abc::def::"``:

.. code-block::

    holder<int>

Output with ``-cr ".+<(.+)>"``:

.. code-block::

    int

How it works
------------

The implementation of **print/printf** does nothing more than forcing the compiler to generate warnings
depending on the passed arguments. The python tool parses the warnings and converts them back to the actually
C++ arguments and outputs them (standardized or formatted) to stdout or stderr.

So what does *-fpermissive* do and why do we use it?

    -fpermissive

    Downgrade some diagnostics about nonconformant code from errors to warnings. Thus, using -fpermissive will allow
    some nonconforming code to compile.

The nonconformant code we use in in the implementation is:

.. code-block:: cpp

    constexpr bool print(int i, int j) {
        int unused = i << j;
        return true;
    }

    constexpr auto test = print(10, 34);

``10 << 34`` will cause an integer overflow which is not allowed, especially in a constant expression.
GCC will output the following interesting diagnostic error:

    <source>:2:20: error: right operand of shift expression '(10 << 34)' is greater than or equal to the precision 32
    of the left operand [-fpermissive]

GCC evaluates the expression ``i << j`` and gives a detailed message about the value of ``i`` and ``j``.
Moreover, the error will recur, even for the same input. Let us all thank GCC for supporting old broken legacy code.
With *-fpermissive* this error becomes a warning and we can `continue compiling <https://gcc.godbolt.org/z/3G8h7M>`__.

So everything we like to print at compile-time and can be broken down to fundamental types, can be outputted.

Is it undefined behavior? Certainly. Will it format erase your hard drive? Probably not.

Use it only for development and not in production!

Known limitations
-----------------

Compiler
++++++++

Since GCC is the only compiler I am aware of with detailed diagnostic warnings to recur, this tool can only work with
GCC. `Prove me wrong. <https://github.com/Viatorus/compile-time-printer/issues/new>`__

Instantiation of static_assert or template parameter
++++++++++++++++++++++++++++++++++++++++++++++++++++

If a CTP statement is used while instantiate an expression triggered by a ``static_assert`` or a `template parameter`,
the compilation will fail without a meaningful error message:

.. code-block::

    <source>:55:19: error: non-constant condition for static assertion
        55 | static_assert(test());
           |               ~~~~^~

Because *-fpermissive* is a legacy option, it is not fully maintained anymore to work across all compile-time
instantiation.

One workaround is to forward the expression to a constexpr variable instantiation:

.. code-block:: cpp

    static_assert(ctp::forward<func()>);

Check out this `example <https://git.io/JLhaX>`__.

Caching
+++++++

The result of a constexpr functions could get cached. If this happens, a CTP statement will only be evaluated once.
Try to generate additional noise to prevent this. Especially if this happens in unevaluated context.
Add additional changing input to the function call as (template) parameter. Also, GCC >=10 added
``-fconstexpr-cache-depth=8``. Maybe a smaller value solves the issue.

Check out this `example <https://git.io/JLhVT>`__.

License
-------

`BSD-1 LICENSE <https://github.com/viatorus/compile-time-printer/blob/develop/LICENSE.txt>`__


CHANGELOG.rst
=========
Changelog
=========

Version 1.0.0
=============

- Release


web/src/js/playground.js
/* global monaco */
import { compile_and_parse } from './ctp';
import CTP_EXAMPLE from 'CTP/tests/data/example.cpp';
import LZString from 'lz-string';

const COMPILERS = [
  ['GCC trunk', 'gsnapshot'],
  ['GCC 10.2', 'g102'],
  ['GCC 10.1', 'g101'],
  ['GCC 9.3', 'g93'],
  ['GCC 9.2', 'g92'],
  ['GCC 9.1', 'g91'],
  ['GCC 8.3', 'g83'],
  ['GCC 8.2', 'g82'],
  ['GCC 8.1', 'g81'],
  ['GCC 7.5', 'g75'],
  ['GCC 7.4', 'g74']
  // ['GCC 7.3', 'g73'], Problem with std::string_view
  // ['GCC 7.2', 'g72']
];

const COMPILER_MESSAGE_RE = /<source>:(\d+):(\d+): ((?:fatal )?error|warning):\s+(.+)/;
const DEFAULT_PLAYGROUND_DATA = {
  code: CTP_EXAMPLE,
  compiler: 'g102',
  compiler_flags: '-fpermissive -std=c++17',
  show_compiler_log: true
};

export class Playground {
  constructor (editor_model) {
    this._data = { ...DEFAULT_PLAYGROUND_DATA };
    this._compiling = null;
    this._output_is_compiling = null;

    this._model = editor_model;
    this._on_change = null;

    // Setup compiler options.
    this._compiler_picker_select = document.getElementById('compiler-picker');
    for (const compiler of COMPILERS) {
      this._compiler_picker_select.options[this._compiler_picker_select.options.length] = new Option(compiler[0], compiler[1]);
    }
    this._compiler_status = document.getElementById('compiler-status');
    this._compiler_flags_input = document.getElementById('compiler-flags');
    this._show_compiler_log_input = document.getElementById('show-compiler-log-input');
    this._output_node = document.getElementById('output');

    this._model.onDidChangeContent(() => {
      this._data.code = this._model.getValue();
      this.compile();
    });
    this._compiler_picker_select.addEventListener('change', e => {
      this._data.compiler = e.target.value;
      this.compile();
    });
    this._compiler_flags_input.addEventListener('input', e => {
      this._data.compiler_flags = e.target.value;
      this.compile();
    });
    this._show_compiler_log_input.addEventListener('change', e => {
      this._data.show_compiler_log = e.target.checked;
      this.compile();
    });
  }

  serialize () {
    return this._data;
  }

  deserialize (data) {
    if (!data) {
      return;
    }
    for (const key of Object.keys(this._data)) {
      if (Object.prototype.hasOwnProperty.call(data, key)) {
        if (typeof this._data[key] === 'boolean') {
          this._data[key] = data[key] === 'true';
        } else {
          this._data[key] = data[key];
        }
      }
    }
    this._update_ui();
  }

  reset () {
    this._data = { ...DEFAULT_PLAYGROUND_DATA };
    this._update_ui();
  }

  register_on_change (callback) {
    this._on_change = callback;
  }

  _update_ui () {
    this._model.setValue(this._data.code);
    for (const option of this._compiler_picker_select.options) {
      if (option.value === this._data.compiler) {
        option.selected = 'selected';
      }
    }
    this._compiler_flags_input.value = this._data.compiler_flags;
    this._show_compiler_log_input.checked = this._data.show_compiler_log;
  }

  compile () {
    if (this._on_change) {
      this._on_change();
    }

    clearTimeout(this._output_is_compiling);
    this._output_is_compiling = setTimeout(() => {
      this._output_node.innerHTML = '';
      const para = document.createElement('span');
      const node = document.createTextNode('<Compiling...>');
      para.classList.add('info');
      para.appendChild(node);
      this._output_node.appendChild(para);

      // Spin compiler status.
      this._compiler_status.style.color = null;
      this._compiler_status.classList.remove('fa-check-circle');
      this._compiler_status.classList.add('fa-spinner');
    }, 500);

    if (this._compiling !== null) {
      this._compiling.cancel();
    }
    this._compiling = compile_and_parse(this._data.compiler, this._data.compiler_flags, this._data.show_compiler_log,
      this._data.code);
    this._compiling.promise.then(([succeeded, printers]) => this._output(succeeded, printers));
  }

  _output (compilation_succeeded, printers) {
    clearTimeout(this._output_is_compiling);

    this._output_node.innerHTML = '';
    let compile_warning = false;
    const widgets = [];
    for (const printer of printers) {
      const para = document.createElement('span');
      const node = document.createTextNode(printer.message);
      if (printer.error_output) {
        para.classList.add('stderr');
      } else {
        para.classList.add('stdout');
      }
      para.appendChild(node);
      this._output_node.appendChild(para);

      if (printer.compiler_output) {
        compile_warning = true;
        const match = COMPILER_MESSAGE_RE.exec(printer.message);
        if (match) {
          const line = parseInt(match[1]);
          if (line > this._model.getLineCount()) {
            continue;
          }
          const start = this._model.getLineFirstNonWhitespaceColumn(line);
          const end = this._model.getLineLastNonWhitespaceColumn(line);
          const severity = match[3] === 'warning' ? 2 : 3;
          widgets.push(
            {
              severity: severity,
              source: match[4],
              message: match[3],
              startLineNumber: line,
              startColumn: start,
              endLineNumber: line,
              endColumn: end
            }
          );
        }
      }
    }
    // Set compiler status.
    this._compiler_status.style.color = null;
    if (compilation_succeeded) {
      this._compiler_status.classList.add('fa-check-circle');
    } else {
      this._compiler_status.classList.add('fa-times-circle');
    }
    if (compile_warning || !compilation_succeeded) {
      this._compiler_status.style.color = 'rgb(255, 101, 0)';
    } else {
      this._compiler_status.style.color = 'rgb(18, 187, 18)';
    }
    this._compiler_status.classList.remove('fa-spinner');

    monaco.editor.setModelMarkers(this._model, 'compilerId', widgets);
  }
}

export class PlaygroundControl {
  constructor (playground) {
    this._playground = playground;
    this._save_playground = true;
    this._short_link = true;

    const reset_btn = document.getElementById('reset-btn');
    reset_btn.addEventListener('click', e => {
      this._playground.reset();
    });

    this._share_link = document.getElementById('share-link');
    document.getElementById('share-link-to-clipboard-btn').addEventListener('click', e => {
      navigator.clipboard.writeText(this._share_link.value);
    });

    document.getElementById('share-link-length').addEventListener('change', e => {
      this._short_link = e.target.value === 'Short';
      this._display_share_link();
    });

    // Open/close share panel.
    const share_panel = document.getElementById('share-panel');
    const open_share_panel_btn = document.getElementById('open-share-panel-btn');
    share_panel.style.top = (open_share_panel_btn.offsetHeight + 5) + 'px';
    open_share_panel_btn.addEventListener('click', () => {
      if (share_panel.style.visibility === 'visible') {
        share_panel.style.visibility = 'hidden';
      } else {
        share_panel.style.visibility = 'visible';
        this._display_share_link();
      }
    });
    window.addEventListener('click', (e) => {
      if (!share_panel.contains(e.target) && !open_share_panel_btn.contains(e.target)) {
        share_panel.style.visibility = 'hidden';
      }
    });

    // Align buttons width and height.
    const github_btn = document.getElementById('github');
    const max_w = Math.max(github_btn.offsetWidth, open_share_panel_btn.offsetWidth, reset_btn.offsetWidth) + 'px';
    github_btn.style.width = max_w;
    reset_btn.style.width = max_w;
    open_share_panel_btn.style.width = max_w;
    document.getElementById('header').style.height = document.getElementById('compiler-picker').offsetHeight + 'px';

    this._playground.register_on_change(() => this._save_to_storage());
  }

  init () {
    let data = this._decode_share_url(window.location.href);
    if (!data) {
      data = this._load_from_storage();
    } else {
      this._save_playground = false;
    }
    this._playground.deserialize(data);
  }

  _load_from_storage () {
    const data = {};
    for (const key of Object.keys(DEFAULT_PLAYGROUND_DATA)) {
      const value = window.localStorage.getItem(key);
      if (value !== null) {
        data[key] = value;
      }
    }
    return data;
  }

  _save_to_storage () {
    if (this._save_playground) {
      const data = this._playground.serialize();
      for (const key of Object.keys(data)) {
        window.localStorage.setItem(key, data[key]);
      }
    }
  }

  _display_share_link () {
    const link = this._create_share_link();
    const url = location.protocol + '//' + location.host + location.pathname + link;
    if (this._short_link) {
      fetch('https://viatorus.pythonanywhere.com/short', {
        method: 'POST',
        body: url
      })
        .then(e => {
          if (e.ok) {
            return e.text();
          }
          return 'Failed to load.';
        })
        .then(txt => { this._share_link.value = txt; });
      this._share_link.value = '';
    } else {
      this._share_link.value = url;
    }
  }

  _create_share_link () {
    const params = [];
    for (const [key, value] of Object.entries(this._playground.serialize())) {
      params.push(`${key}=${encodeURIComponent(value)}`);
    }
    return '#' + LZString.compressToBase64(params.join('&'));
  }

  _decode_share_url (url) {
    const i = url.indexOf('#');
    if (i >= 0) {
      const decoded = LZString.decompressFromBase64(url.slice(i + 1));
      const data = {};
      for (const param of decoded.split('&')) {
        const [name, value] = param.split('=');
        if (typeof value === 'string') {
          data[name] = decodeURIComponent(value);
        }
      }
      return data;
    }
    return null;
  }
}
